criu/fsnotify.c

```diff
@@ -228,63 +231,59 @@ int check_open_handle(unsigned int s_dev, unsigned long i_ino,
 		FhEntry *f_handle)
 {
 	char *path, *irmap_path;
-	int fd = -1;
+	struct mount_info *mi;
 
-	if (fault_injected(FI_CHECK_OPEN_HANDLE)) {
-		fd = -1;
+	if (fault_injected(FI_CHECK_OPEN_HANDLE))
 		goto fault;
-	}
 
-	fd = open_handle(s_dev, i_ino, f_handle);
-fault:
-	if (fd >= 0) {
-		struct mount_info *mi;
+	/*
+	 * Always try to fetch watchee path first. There are several reasons:
+	 *
+	 *  - tmpfs/devtmps do not save inode numbers between mounts,
+	 *    so it is critical to have the complete path under our
+	 *    hands for restore purpose;
+	 *
+	 *  - in case of migration the inodes might be changed as well
+	 *    so the only portable solution is to carry the whole path
+	 *    to the watchee inside image.
+	 */
+	path = alloc_openable(s_dev, i_ino, f_handle);
 
+	if (!IS_ERR_OR_NULL(path)) {
 		pr_debug("\tHandle 0x%x:0x%lx is openable\n", s_dev, i_ino);
+		goto out;
+	} else if (IS_ERR(path) && PTR_ERR(path) == -3) {
```

Comment:
You need to define constants for these error codes and describe them in the comment for alloc_openable.

Reply:
I have added constants for error codes. They look self-explaining to me, but I can add additional comment if you want.